draw code minings at end of line before or after the eol character

tobiasmelcher opened the pull request:
We would like to implement a ghost text like completion functionality like in vscode and want to re-use the existing eclipse code mining implementation.  

It works quite well as long the completion trigger offset is not the last character of the line.
Following screenshot shows a scenario where the user has pressed the character "e" and the keyword "entity" is then proposed. The code mining consists of the rest of the word "ntity".
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/10373336/2ab793b7-1856-4eea-a9b0-efe186fbdda3)

If the cursor is at the end of the line, the code mining is shown with some space next to the (invisible) line break character which does not look that nice.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/10373336/57d1a7ff-177f-45c0-9f0d-1b55b8f59cfd)

This change introduces method
AbstractCodeMining#miningToBeDrawnAfterLineBreakCharacterIfAtEndOfLine so that mining implementors can decide where the code mining should be drawn: before or after the (invisible) end of line character.

Here a screenshot where miningToBeDrawnAfterLineBreakCharacterIfAtEndOfLine returns false.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/10373336/601c5da5-314f-4166-a7d6-f4e733f49b57)

What do you think? Does this change make sense? Do you see any issues?

tobias-melcher committed:
draw code minings at end of line before eol character if possible

mickaelistria commented:
The behavior you suggest would look like the one anyone would expect in such case, why making a new API option for it. IMO, it's expected that if the character which is decorated is a newline, then code mining would be printed before it.
Unless I'm missing an important use-case where drawing after the newline character is functionally important (the only usage I know is debug values at end of line, but the newline character position doesn't matter for it), I suggest you just make the proposed behavior default without changing the API.

BeckerWdf commented:
> I suggest you just make the proposed behavior default without changing the API.

I also think this should be possible. But we saw that this behaviour was introduced with 
https://git.eclipse.org/r/c/platform/eclipse.platform.text/+/190503/3/org.eclipse.jface.text/src/org/eclipse/jface/text/source/inlined/InlinedAnnotationDrawingStrategy.java by @mickaelistria .
See also the discussion in https://bugs.eclipse.org/bugs/show_bug.cgi?id=540448#c30

We didn't quite understand the discussion and the use-case there so @tobiasmelcher proposed this with an API.

@mickaelistria: If you can explain why you did this change back then we maybe find out what would break if we now simply make the proposed behavior default without changing the API. Or we come to the conclusion that we really don't break anything with this.

mickaelistria commented:
The change back then was to support drawing an code mining at a end of line. IIRC, the particularity of line end is that there is no character printed, so the usual trick with StyledText range to widen the character area and draw extra content in the area wasn't working.
But it could be that we implemented a bunch of changes since then, such as drawing to the right of previous char by default which make that this particular case of line end can be ditched to get better results with less code?
Anyway, feel free to change anything, as long as it works better, I'm fine with it (I'm not so emotionally attached to the code I write, I'm more attached to the resulting feature, so whatever makes the feature better makes me happier, even if it removes tons of my code)

tobiasmelcher commented:
> The change back then was to support drawing an code mining at a end of line. IIRC, the particularity of line end is that there is no character printed, so the usual trick with StyledText range to widen the character area and draw extra content in the area wasn't working. But it could be that we implemented a bunch of changes since then, such as drawing to the right of previous char by default which make that this particular case of line end can be ditched to get better results with less code? Anyway, feel free to change anything, as long as it works better, I'm fine with it (I'm not so emotionally attached to the code I write, I'm more attached to the resulting feature, so whatever makes the feature better makes me happier, even if it removes tons of my code)

Thanks a lot for the feedback. Please check my latest change in which I removed the API. The minings are now drawn in front of the line break character when possible. It is not possible when the line is empty, I therefore couldn't remove the call drawAfterLine(annotation, gc, textWidget, widgetOffset, length, color);.
What do you think?

mickaelistria commented:
Can you please check how it behaves with Preferences > Run/Debug > Show debug values inline on code editors turned on and verify it doesn't break it? Make sure you're using the smallest project and file, otherwise this feature can bloat CPU.

BeckerWdf commented:
> Can you please check how it behaves with Preferences > Run/Debug > Show debug values inline on code editors turned on and verify it doesn't break it? Make sure you're using the smallest project and file, otherwise this feature can bloat CPU.

Hi Mickael,

I tried to run a simple hello world java program and don't see the annotations even though I have enabled them.
In https://eclipse.dev/eclipse/news/4.23/platform.php#inline-debug-values I read:

> This is an experimental feature and since this requires extra effort for the debuggers to enable it, not all debuggers may support it yet e.g. **Java debugger does not support it yet**.

Is this the cause?

How can I easily test this?

mickaelistria commented:
I tested it with debugging a .js file and while the rendering of debugger info is slightly worse (some prefix space is lost), the behavior seems more consistent and the debug code-mining provide should just be tweaked to include the desired spacing in the items it displays.
Thanks for this good patch, and for sharing the interesting use-case you're building on top of it!

BeckerWdf commented:
> I tested it with debugging a .js file and while the rendering of debugger info is slightly worse (some prefix space is lost), the behavior seems more consistent and the debug code-mining provide should just be tweaked to include the desired spacing in the items it displays. Thanks for this good patch, and for sharing the interesting use-case you're building on top of it!

Can you show a before and after screenshot?

mickaelistria commented:
before
![Screenshot from 2024-06-20 09-10-47](https://github.com/eclipse-platform/eclipse.platform.ui/assets/469058/2739822e-ed6b-4544-afb4-dfb377d72589)
After
![Screenshot from 2024-06-20 09-11-58](https://github.com/eclipse-platform/eclipse.platform.ui/assets/469058/59432446-48ea-421e-9bab-b0af29a0ac6d)

While it doesn't look as good for this case, the "after" behavior is more correct and more powerful. It's up to the code-mining provider to deal with extra spacing.
